src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +243,190 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+                }
+                else
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
+                    return null;
+                }
+            }
+            else if (serverNamePartsByComma.Length < 2 && serverNamePartsByFowardSlash.Length == 2)
+            {
+                string hostName = serverNamePartsByFowardSlash[0];
+                string instanceName = serverNamePartsByFowardSlash[1];
+                int portNumber = -1;
+
+                if (GetPortByInstanceName(hostName, instanceName, out portNumber))
                 {
-                    portNumber = ushort.Parse(serverAndPortParts[1]);
+                    return new SNITCPHandle(hostName, portNumber, timerExpire, callbackObject, parallel);
                 }
-                catch (Exception e)
+                else
                 {
-                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                     return null;
                 }
             }
-            else if (serverAndPortParts.Length > 2)
+            else
             {
                 SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                 return null;
             }
+        }
+
+        /// <summary>
+        /// Finds IPAddress by given hostname.
+        /// </summary>
+        /// <param name="hostname">hostname string</param>
+        /// <param name="hostIPAddress">host IPAddress found by hostname string</param>
+        /// <param name="family">address family. Default value is AddressFamily.InterNetwork</param>
+        /// <returns>bool value showing if operation is successful</returns>
+        private static bool GetIPAdress(string hostname, out IPAddress hostIPAddress, AddressFamily family = AddressFamily.InterNetwork)
+        {
+            hostIPAddress = null;
+
+            if (String.IsNullOrWhiteSpace(hostname))
+            {
+                return false;
+            }
+
+            try
+            {
+                Task<IPAddress[]> addressTask = Dns.GetHostAddressesAsync(hostname);
+                IPAddress[] addresses = null;
+                if (!addressTask.Wait(1000) || (addresses = addressTask.Result) == null || addresses.Length == 0)
+                {
+                    return false;
+                }
+
+                List<IPAddress> addressList = new List<IPAddress>(addresses);
+                hostIPAddress = addressList.First(a => a.AddressFamily == family);
+                if (hostIPAddress == null)
```

Comment:
Why allocate a new `List<IPAddress>` and copy `addresses`?

This could be simplified as follows, which doesn't have any unnecessary allocations, and avoids having to catch the exception thrown from `First` when the item isn't found:

```c#
foreach (IPAddress address in addresses)
{
    if (address.AddressFamily == family)
    {
        hostIPAddress = address;
        return true;
    }
}

return false;
```